synapse/storage/databases/main/__init__.py

```diff
@@ -138,6 +145,19 @@ def __init__(
 
         super().__init__(database, db_conn, hs)
 
+        def repl(match: Match[str]) -> str:
+            return "_" + match.group(0).lower()
+
+        for datastore_class in self.DATASTORE_CLASSES:
+            name = datastore_class.__name__
+            if name.endswith("Store"):
+                name = name[: -len("Store")]
+
+            name = re.sub(r"[A-Z]", repl, name)[1:]
+
+            store = datastore_class(database, db_conn, hs, self)
+            setattr(self, name, store)
```

Comment:
I'd be tempted to avoid the magic here and suffer the boilerplate of 

```
self.relations = RelationsStore(db_conn, hs, self)
```

to avoid the cleverness.

Or perhaps my point is: I want to avoid dynamic lists of stores where possible

Reply:
That's reasonable since right now we define them twice on the class anyway, might be easier to just list them out.

Reply:
Unfortunately `mypy` still seems to need the attribute to be defined on the class or it doesn't know how to resolve `relations`. With the following diff:

```diff
diff --git a/synapse/storage/databases/main/__init__.py b/synapse/storage/databases/main/__init__.py
--- a/synapse/storage/databases/main/__init__.py
+++ b/synapse/storage/databases/main/__init__.py
@@ -128,13 +128,6 @@ class DataStore(
     LockStore,
     SessionStore,
 ):
-    DATASTORE_CLASSES: List[Type[SQLBaseStore]] = [
-        RelationsStore,
-    ]
-
-    # XXX So mypy knows about dynamic properties.
-    relations: RelationsStore
-
     def __init__(
         self,
         database: DatabasePool,
@@ -147,18 +140,8 @@ class DataStore(
 
         super().__init__(database, db_conn, hs)
 
-        def repl(match: Match[str]) -> str:
-            return "_" + match.group(0).lower()
-
-        for datastore_class in self.DATASTORE_CLASSES:
-            name = datastore_class.__name__
-            if name.endswith("Store"):
-                name = name[: -len("Store")]
-
-            name = re.sub(r"[A-Z]", repl, name)[1:]
-
-            store = datastore_class(database, db_conn, hs, self)
-            setattr(self, name, store)
+        # This is a bit repetitive, but avoids dynamically setting attributes.
+        self.relations = RelationsStore(database, db_conn, hs, self)
 
     async def get_users(self) -> List[JsonDict]:
         """Function to retrieve a list of users in users table.
```

When running mypy it still squeaks (note I added a couple of `reveal_type` calls in there to help understand what's happening):

```
$ mypy
synapse/api/filtering.py:511: note: Revealed type is "synapse.storage.databases.main.DataStore"
synapse/api/filtering.py:513: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2024: note: Revealed type is "synapse.storage.databases.main.DataStore"
synapse/storage/databases/main/events.py:2026: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2031: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2037: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2042: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2046: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2050: error: Cannot determine type of "relations"  [has-type]
synapse/storage/databases/main/events.py:2054: error: Cannot determine type of "relations"  [has-type]
Found 8 errors in 2 files (checked 789 source files)
```

----

Any thoughts if that is solvable?

Reply:
From https://mypy.readthedocs.io/en/stable/error_code_list.html?highlight=cannot%20determine#check-that-type-of-target-is-known-has-type:

> Mypy sometimes generates an error when it hasn’t inferred any type for a variable being referenced. This can happen for references to variables that are initialized later in the source file, and for references across modules that form an import cycle. When this happens, the reference gets an implicit Any type.

I would guess that something like this is happening here. (To define a DataStore, you need a RelationsStore...but that also takes a DataStore. It feels circular.)

Adding back `relations: RelationsStore` in the class body fixes this though, as does writing `self.relations: RelationsStore = RelationsStore(...)` in the initialiser.

Reply:
> Adding back `relations: RelationsStore` in the class body fixes this though,

Yep, but the goal was to *not* list them all twice.

> as does writing `self.relations: RelationsStore = RelationsStore(...)` in the initialiser.

This seems annoying, but a better solution?